runtime/runtime/src/lib.rs

```diff
@@ -319,12 +403,13 @@ impl Runtime {
                     block_index,
                     nonce.as_ref().to_vec(),
                     false,
+                    public_key,
                 ),
             )
             .map_err(|e| format!("wasm async call preparation failed with error: {:?}", e))?;
```

Comment:
`executor::execute` returns `Err` also when the contract code fails (e.g. due to running out of liquid balance) not only when preparation fails. We should change the error message here.

Reply:
This error should only be a preparation error. Since we need to keep track of the liquid balance in case of contract failure (for whatever reason).

Reply:
We return `Ok` from `execute` in case the contract run out of liquid balance

Reply:
So when `charge_balance_with_limit` returns `Err` then `execute` returns `Ok`?

Reply:
Yes, but it includes this error within `.return_data`